### PR TITLE
Port LDAP Authenticator to python 3

### PR DESCRIPTION
**Pull request by andynd**
As the title says

### CODE REVIEW

**Comment by andynd** [comment]
relates to #26

**Comment by Kissaki** [comment]
Do you use this productively? What was your test approach?

File: Authenticators/LDAP/LDAPauth.py [hunk 1]
```diff
@@ -97,18 +98,24 @@
 # authentication scheme.
 #
 #    Requirements:
-#        * python >=2.4 and the following python modules:
+#        * python >=3.8 (maybe 3.6 is enought but it wasn't tested) and the following python modules:
```
Comment (Kissaki) [inline Authenticators/LDAP/LDAPauth.py:101]:
```suggestion
#        * python >=3.8 (maybe 3.6 is enough but it wasn't tested) and the following python modules:
```

File: Authenticators/LDAP/LDAPauth.py [hunk 2]
```diff
@@ -97,18 +98,24 @@
...
 #            * ice-python
 #            * python-ldap
 #            * daemon (when run as a daemon)
+#    If you are using Ubuntu/Debian (only Ubuntu 20.04 was tested) this means the following packages:
```
Comment (Kissaki) [inline Authenticators/LDAP/LDAPauth.py:105]:
```suggestion
#    If you are using Ubuntu/Debian (only Ubuntu 20.04 was tested) the following packages provide these:
```

File: Authenticators/LDAP/LDAPauth.py [hunk 3]
```diff
@@ -97,18 +98,24 @@
...
+#        * python3-ldap
+#        * python3-daemon
+#        * zeroc-ice-slice
 
 import sys
 import ldap
 import Ice
-import thread
-import urllib2
+import _thread
```
Comment (Kissaki) [inline Authenticators/LDAP/LDAPauth.py:115]: What’s the significance of the underscore prefix? What does it mean?

File: Authenticators/LDAP/LDAPauth.py [hunk 4]
```diff
@@ -514,7 +521,7 @@ def authenticate(self, name, pw, certlist, certhash, strong, current = None):
                 
             # Parse the user information.
             uid = int(match[1][cfg.ldap.number_attr][0])
-            displayName = match[1][cfg.ldap.display_attr][0]
+            displayName = match[1][cfg.ldap.display_attr][0].decode()
```
Comment (Kissaki) [inline Authenticators/LDAP/LDAPauth.py:524]: What does this decode from and into? What’s the context to this?

Comment (andynd) [reply Authenticators/LDAP/LDAPauth.py:524]: ldap returns bytes, we need a string

File: Authenticators/LDAP/LDAPauth.py (hunk 3 above)
Comment (andynd) [reply Authenticators/LDAP/LDAPauth.py:115]: It's the module name: https://docs.python.org/3/library/_thread.html

**Comment by andynd** [comment]
> Do you use this productively? What was your test approach?

Yes it is used in production for the syseleven internal mumbel server. Testing was a few manual login attemps with working and not working passwords as well as users which are/aren't in the (config defined) group.

File: Authenticators/LDAP/LDAPauth.py (hunk 3 above)
Comment (Kissaki) [reply Authenticators/LDAP/LDAPauth.py:115]: But why do they add an underscore to it?

**Review by Kissaki — APPROVED** [review]

File: Authenticators/LDAP/LDAPauth.py (hunk 3 above)
Comment (andynd) [reply Authenticators/LDAP/LDAPauth.py:115]: Thats a great question to ask the python developer why they name their module like that but not related to this pull request... the python 2 "thread" module is simply named "_thread" in python 3

File: Authenticators/LDAP/LDAPauth.py (hunk 4 above)
Comment (Krzmbrzl) [reply Authenticators/LDAP/LDAPauth.py:524]: shouldn't we specify a code-page that should be used for decoding?

Comment (andynd) [reply Authenticators/LDAP/LDAPauth.py:524]: We could, but it defaults to utf-8 and that will work with any modern ldap server without a very strange config.

Comment (Krzmbrzl) [reply Authenticators/LDAP/LDAPauth.py:524]: Okay then :)

**Review by Krzmbrzl — APPROVED** [review]

**Comment by w1n5l0w** [comment]
@andynd I do not get the script to read the LDAPauth.ini which is placed in the same folder. Neither by running it directly, nor if I run it as systemd service as descriped here: https://github.com/mumble-voip/mumble-scripts/issues/19
It doesn't work as root or as mumble-server (when ownership is set accordingly to the respective user and even permissions are 777). OS ist Ubuntu 20.04 LTS and all python3.8 modules are installed as written in the description.

Error message is `Fatal error, could not load config file from "LDAPauth.ini"`

Any idea what might be missing? The python2 script worked like charm for me, "LDPAPauth.ini hasn't been changed since.

**Comment by Kissaki** [comment]
Please create a new ticket so it is visible as an open ticket/support request and so the comments on this ticket stay on-topic. Other users may have similar issues and not find it here.

**Comment by andynd** [comment]
Hmm, sadly the `try..except` in the code hides the actual exception. Could you comment that out (line 836) and post the actual exception?

**Pull request merged**